misc/profiles2/trekking.brf

```diff
@@ -52,7 +54,147 @@ assign is_ldcr =
      if ignore_cycleroutes then false
      else any_cycleroute
 
-assign isbike = or bicycle=yes or or bicycle=permissive bicycle=designated lcn=yes
+#
+# handle one-ways. On primary roads, wrong-not oneways should
+# be close to forbidden, while on other ways we just add
+# 4 to the costfactor (making it at least 5 - you are allowed
+# to push your bike)
+#
+assign badoneway =  # True if we are going against traffic in a one-way street
+       if reversedirection=yes then
+         if oneway:bicycle=yes then true
+         else if oneway= then junction=roundabout
+         else oneway=yes|true|1
+       else oneway=-1
+
+assign goodoneway = if reversedirection=yes then oneway=-1  # True if we are going along traffic in a one-way street
+                      else if oneway= then junction=roundabout else oneway=yes|true|1
+
+assign onewaypenalty =  # Penalty associated with going against traffic in a one-way stret
+       # Note: Can be zero if opposite traffic is allowed for bikes
+       if ( badoneway ) then
+       (
+         if ( cycleway=opposite|opposite_lane|opposite_track ) then 0
+         # No penalty if there is any cycleway in opposite direction
+         # Note: We are already contraflow, so no need for more checks.
+         else if ( cycleway:left=opposite|opposite_lane|opposite_track ) then 0
+         else if ( cycleway:right=opposite|opposite_lane|opposite_track ) then 0
+         else if ( cycleway:left:oneway=-1|no ) then 0
```

Comment:
Should not this be dependent on reversedirection=yes?
```suggestion
          else if ( and reversedirection=yes cycleway:left:oneway=-1|no ) then 0
          else if ( and not reversedirection=yes cycleway:left:oneway=yes ) then 0
```

Reply:
This is enclosed within a `if ( badoneway) `: https://github.com/abrensch/brouter/pull/123/files#diff-2bbe87cbb0ab11e03987f2ebbd823c31R75

which should test for `reversedirection` https://github.com/abrensch/brouter/pull/123/files#diff-2bbe87cbb0ab11e03987f2ebbd823c31R63.

So, if I'm not mistaking, this part should only consider `cycleway:*=-1|no` which means against traffic (`reversedirection`).

If we are going same way as traffic, there should not be any particular `onewaypenalty` since we are not going against traffic in any way.

Reply:
Example:
badoneway = true because reversedirection=<null> and oneway=-1.
In other words:
- We are going in direction as way was mapped
- direction of way does not agree with direction of one way route. This is discouraged, but can happen.

The one way for cyclist agrees with direction of way in map so cycleway:left:oneway=yes

Is it clear now? Or is it me who get it wrong?

Reply:
Oh, indeed, thanks for the details. I guess this should be fixed in https://github.com/abrensch/brouter/pull/123/commits/5a1f7dccedc0c97f21166f0b661d9f562b8f7829. Could you check please?

Parts of this PR has already been included in https://github.com/abrensch/brouter/pull/136. I'll make smaller (atomic) commits to help merge the rest of this large one afterwards :)

Reply:
I have additional comments, but reading @abrensch comment, might be good to start new pull request and close this one. (Subject to style and opinion)

It might be worth to go through all conditions again and check them. Like line [80](https://github.com/abrensch/brouter/blob/5a1f7dccedc0c97f21166f0b661d9f562b8f7829/misc/profiles2/trekking.brf#L80) was not fixed. Maybe worth refactoring a bit.

Also I do not understand line [133](https://github.com/abrensch/brouter/blob/5a1f7dccedc0c97f21166f0b661d9f562b8f7829/misc/profiles2/trekking.brf#L133).

Reply:
First, thanks for your feedback and review on this! Editing BRouter profiles can be tricky and having mulitple pairs of eyes looking at the code is definitely great!

> I have additional comments, but reading @abrensch comment, might be good to start new pull request and close this one. (Subject to style and opinion)

This is what I had in mind when I started #136 with atomic edits, one edit per commit to ease review and discussion. :) Plus there are some duplicates and introducing intermediate variables would help a lot here I think.

If you have any comments on what is already included in #136, please feel free to review on this one, so this one can get fixed and eventually merged :) Then, I think we can move on to the following edits.

> It might be worth to go through all conditions again and check them.

Sure! Additionnally, if you know any real life examples for these conditions (OSM data), please feel free to report them to me or add them to https://github.com/Phyks/BrouterTesting. I used this to ensure my edits were not completely absurd and working on some features, but I don't have tests for all the conditions.

> Like line 80 was not fixed.

Actually, I was thinking this one was right. The issue here is about the meaning of `cycleway=opposite*` in OSM I think. My understanding was that in a oneway street, `cycleway:*=opposite*` means that the cycleway goes against traffic in a one-way street, which is always what we should be looking for when within a `badoneway`.

> Also I do not understand line 133.

Yes, I guess this should be written the same way as line 80 anyways. This part is quite difficult to read and definitely needs refactoring.